File: backend/blog-service/src/main/java/com/lily/blogservice/client/BlogSearchClient.java
package com.lily.blogservice.client;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.lily.blogservice.dto.BlogDocument;
import com.lily.blogservice.dto.BlogSearchRequest;
import com.lily.blogservice.dto.BlogSearchResult;
import com.lily.blogservice.dto.SearchMeta;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cloud.client.circuitbreaker.CircuitBreaker;
import org.springframework.cloud.client.circuitbreaker.CircuitBreakerFactory;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpMethod;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;


@Slf4j
@Component
@RequiredArgsConstructor
public class BlogSearchClient {

  static private final ObjectMapper MAPPER = new ObjectMapper();

  private final RestTemplate restTemplate;

  private final KakaoBlogSearchSource kakaoBlogSearchSource;

  private final NaverBlogSearchSource naverBlogSearchSource;

  private final CircuitBreakerFactory circuitBreakerFactory;


  private BlogSearchResult _search(final BlogSearchSource source, final BlogSearchRequest request) {
    final String urlTemplate = source.getUrlTemplate();
    final HttpEntity<?> entity = new HttpEntity<>(source.getHttpHeaders());
    final Map<String, Object> parameters = source.getParameters(request);

    final HttpEntity<String> response = restTemplate.exchange(
        urlTemplate,
        HttpMethod.GET,
        entity,
        String.class,
        parameters
    );

    final String jsonString = response.getBody();
    Map<String, Object> json = new HashMap<>();
    try {
      json = MAPPER.readValue(jsonString, Map.class);
    } catch (JsonProcessingException e) {
      log.error("[Todo check] json mapping failed", e);
    }

    final SearchMeta meta = source.parseSearchMeta(json);
    final List<BlogDocument> documents = source.parseBlogDocuments(json);

    return new BlogSearchResult(meta, documents);
  }

  public BlogSearchResult search(final BlogSearchRequest request) {
    CircuitBreaker circuitBreaker = circuitBreakerFactory.create("searchClientCircuitBreaker");
    CircuitBreaker lastCircuitBreaker = circuitBreakerFactory.create("last");

    return circuitBreaker.run(
        () -> _search(kakaoBlogSearchSource, request),
        throwable -> lastCircuitBreaker.run(
            () -> _search(naverBlogSearchSource, request),
            finalThrowable -> BlogSearchResult.emptyResult())
    );
  }

}

File: backend/blog-service/src/main/kotlin/com/lily/blogservice/client/BlogSearchClient.kt
package com.lily.blogservice.client

import com.lily.blogservice.data.BlogSearchRequest
import com.lily.blogservice.data.BlogSearchResult
import kotlinx.coroutines.reactor.awaitSingle
import org.springframework.core.ParameterizedTypeReference
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClient

@Component
class BlogSearchClient(val source: BlogSearchSource, val webClient: WebClient) {

    suspend fun search(request: BlogSearchRequest?): BlogSearchResult {
        val uriTemplate = source.getUrlTemplate()
        val headers = source.getHttpHeaders()
        val parameters = source.getParameters(request!!)

        val json = webClient.get()
            .uri(uriTemplate.expand(parameters))
            .headers { httpHeaders ->
                headers.keys.stream().forEach {
                    httpHeaders.set(it, headers.getValue(it))
                }}
            .retrieve()
            .bodyToMono(object : ParameterizedTypeReference<Map<String, Any>>(){})
            .awaitSingle()// Mono -> Coroutine으로 변환

        val meta = source.parseSearchMeta(json)
        val documents = source.parseBlogDocuments(json)

        return BlogSearchResult(meta, documents)
    }
}